src/algebra/lie_algebras.lean

```diff
@@ -0,0 +1,138 @@
+Authors: Oliver Nash
+-/
+import ring_theory.algebra data.matrix.basic
+
+/-!
+# Lie algebras
+
+This file defines Lie rings, and Lie algebras over a commutative ring. It shows how these arise from
+associative rings and algebras via the ring commutator. In particular it defines the Lie algebra
+of endomorphisms of a module as well as of the algebra of square matrices over a commutative ring.
+
+## Notations
+
+We introduce the notation ⦋x, y⦌ for the Lie bracket. Note that these are the Unicode "square
+brackets with underbar" rather than the usual square brackets.
+
+## Implementation notes
+
+Lie algebras are defined as modules with a compatible Lie ring structure, and thus are partially
+unbundled. Since they extend Lie rings, these are also partially unbundled.
+
+## References
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 1--3*][bourbaki1975]
+
+## Tags
+
+lie ring, lie algebra, lie bracket, jacobi identity, ring commutator
+-/
+
+universes u v
+
+class has_bracket (L : Type v) := (bracket : L → L → L)
+
+notation `⦋`x`,` y`⦌` := has_bracket.bracket x y
+
+/--
+A Lie ring is an additive group with compatible product, known as the bracket, satisfying the
+Jacobi identity. The bracket is not associative unless it is identically zero.
+-/
+class lie_ring (L : Type v) [add_comm_group L] extends has_bracket L :=
+(add_left : ∀ (x y z : L), ⦋x + y, z⦌ = ⦋x, z⦌ + ⦋y, z⦌)
+(add_right : ∀ (x y z : L), ⦋z, x + y⦌ = ⦋z, x⦌ + ⦋z, y⦌)
+(alternate : ∀ (x : L), ⦋x, x⦌ = 0)
+(jacobi : ∀ (x y z : L), ⦋x, ⦋y, z⦌⦌ + ⦋y, ⦋z, x⦌⦌ + ⦋z, ⦋x, y⦌⦌ = 0)
+
+/--
+A Lie algebra is a module with compatible product, known as the bracket, satisfying the Jacobi
+identity. Forgetting the scalar multiplication, every Lie algebra is a Lie ring.
+-/
+class lie_algebra (R : Type u) (L : Type v)
+  [comm_ring R] [add_comm_group L] extends module R L, lie_ring L :=
+(smul_right : ∀ (t : R) (x y : L), ⦋x, t • y⦌ = t • ⦋x, y⦌)
+
+namespace lie_algebras
+
+  lemma lie_ring_skew (L : Type v) [add_comm_group L] [lie_ring L] (x y : L) :
+    ⦋x, y⦌ = -⦋y, x⦌ :=
+  begin
+    rw [←sub_eq_zero_iff_eq, sub_neg_eq_add],
+    have h : ⦋x + y, x + y⦌ = 0 := by apply lie_ring.alternate,
+    rw [lie_ring.add_left,
+        lie_ring.add_right, lie_ring.add_right,
+        lie_ring.alternate, lie_ring.alternate] at h,
+    simp at h,
+    exact h,
+  end
+
+  lemma lie_algebra_smul_left (R : Type u) (L : Type v)
+    [comm_ring R] [add_comm_group L] [lie_algebra R L] (t : R) (x y : L) :
+    ⦋t • x, y⦌ = t • ⦋x, y⦌ :=
+  by rw [lie_ring_skew, lie_ring_skew _ x y, lie_algebra.smul_right]; simp
+
+  def adjoint_action (R : Type u) (L : Type v)
+    [comm_ring R] [add_comm_group L] [lie_algebra R L] (x : L) : L →ₗ[R] L :=
+  { to_fun := has_bracket.bracket x,
+    add    := by intros; apply lie_ring.add_right,
+    smul   := by intros; apply lie_algebra.smul_right }
+
+  def lie_bracket_bilinear (R : Type u) (L : Type v)
```

Comment:
I wonder if it would be useful to have the bracket be definitionally this bilinear map.

Reply:
Possibly, and I did consider this. The reason I went for what I have was that otherwise the only way we can define `lie_ring` is to regard it formally as a ℤ-module which seems rather heavyweight.

Still, it's a subjective call and it would not take much to persuade me to make the bilinear map the definition.

Reply:
Fair enough. Let's leave it like this for now.